---
title: "PING (Pushing Immunogenetics to the Next Generation)- set the copy number thresholds and run PING2 alignments and allele calling"
author: "Jieqiong Dai"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
      html_document:
        code_folding: hide
        toc: true
        toc_float: true
        pandoc_args: [
         "+RTS", "-k16g",
         "-RTS"
         ]   
params:
      data: ''

---

<style type="text/css">

body{ /* Normal  */
      font-size: 12px;
  }
td {  /* Table  */
  font-size: 12px;
}
h1.title {
  font-size: 38px;
  color: DarkRed;
}
h1 { /* Header 1 */
  font-size: 20px;
  color: DarkBlue;
}
h2 { /* Header 2 */
  font-size: 18px;
  color: DarkBlue;
}
h3 { /* Header 3 */
  font-size: 16px;
  color: DarkBlue;
}
code.r{ /* Code block */
    font-size: 12px;
}
pre { /* Code block - determines code spacing between lines */
    font-size: 14px;
}
</style>


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
# Working directory:
```{r,message=FALSE,warning=FALSE}
print(getwd())
```
# Check required packages
```{r,message=FALSE,warning=FALSE}
# function for installing needed packages
installpkg <- function(x){
    r = getOption("repos")
    r["CRAN"] = "http://cran.us.r-project.org"
    options(repos = r)
    if(x %in% rownames(installed.packages())==FALSE) {
        if(x %in% rownames(available.packages())==FALSE) {
            if (!requireNamespace("BiocManager", quietly = TRUE))
               install.packages("BiocManager")
               BiocManager::install(x,ask = FALSE)
        } else {
            install.packages(x)
        }
        paste(x, "package is installed...")
    } else {
        paste(x,"package already installed...")
    }
}

# install necessary packages
required_packages  <- c("data.table","plotly","stringr","pryr","gtools","R.utils","rmarkdown","methods","zip","BiocParallel")
lapply(required_packages,installpkg)

library(data.table)
library(stringr)
library(methods)
library(pryr)
library(plotly)
library(R.utils)
library(gtools)
library(zip)
```
```{r setup, include=FALSE}
#paralle
library(BiocParallel)
# code from https://hpc.nih.gov/apps/R.html#biocparallel
detectBatchCPUs <- function() {
    ncores <- as.integer(Sys.getenv("SLURM_CPUS_PER_TASK"))
    if (is.na(ncores)) {
        ncores <- as.integer(Sys.getenv("SLURM_JOB_CPUS_PER_NODE"))
    }
        if (is.na(ncores)) {
                return(4) # for helix
        }
        return(ncores)
}
ncpus <- detectBatchCPUs()
register(MulticoreParam(ncpus))
```
# Set up system and parameters
```{r,message=FALSE,warning=FALSE}
load(params$data)
source('Resources/general_functions.R')
source('Resources/extractor_functions.R') 
source('Resources/ping_copy.R') 
source('Resources/ping_allele.R') 
source('Resources/ping_gc_align.R') 
source('Resources/alleleCombine_functions.R') 

threads <- ncpus
setDTthreads(threads)
```

# Set copy number thresholds
```{r,message=FALSE,warning=FALSE}
sampleList <- ping_copy.manual_threshold(sampleList=sampleList,resultsDirectory=outDir$path,use.threshFile = T) # this function sets copy thresholds
sampleList <- ping_copy.load_copy_results(sampleList, outDir$path)

```

# PING2 alignments and allele calling
```{r,message=FALSE,warning=FALSE}
if(allele.fullAlign){
  as.list <- alleleSeq.list
}else{
  as.list <- compact.alleleSeq.list
}

probelistFile='probelist_2021_01_24.csv'
gcResourceDirectory <- normalizePath('Resources/gc_resources', mustWork = T)
cat('\n\nReading in the KFF probelist file: ', file.path(gcResourceDirectory, probelistFile))
probeDF <- read.csv(file.path(gcResourceDirectory, probelistFile), stringsAsFactors = F, check.names = F)
row.names(probeDF) <- probeDF$Name

# Alignment and allele calling workflow
sampleList <- ping_allele(sampleList)

# ----- Formatting Results Genotypes -----
source('Resources/alleleFinalize_functions.R')
#cat('\n\n ----- FINALIZING GENOTYPES ----- ')
finalCallPath <- pingFinalize.format_calls( resultsDirectory )
cat('\nFinal calls written to:',finalCallPath)

store2 <- paste0(resultsDirectory,"/ping3.Rdata")
save.image(store2)
```